modules/augain/augain.c

```diff
@@ -0,0 +1,161 @@
+/**
+ * @file augain.c  Audio gain
+ *
+ * Copyright (C) 2024 Juha Heinanen
+ */
+
+#include <stdlib.h>
+#include <re.h>
+#include <rem.h>
+#include <baresip.h>
+
+
+/**
+ * Audio gain module
+ *
+ * This module can be used to increase volume of audio source,
+ * for example, microphone.
+ *
+ * Sample config:
+ *
+ \verbatim
+  augain            1.5
+ \endverbatim
+ */
+
+
+struct augain_enc {
+	struct aufilt_enc_st af;  /* base class */
+};
+
+
+static double gain = 1.0;
+
+
+static void enc_destructor(void *arg)
+{
+	struct augain_enc *st = arg;
+
+	list_unlink(&st->af.le);
+}
+
+
+static int encode_update(struct aufilt_enc_st **stp, void **ctx,
+			 const struct aufilt *af, struct aufilt_prm *prm,
+			 const struct audio *au)
+{
+	struct augain_enc *st;
+	(void)ctx;
+	(void)af;
+	(void)au;
+
+	if (!stp || !prm)
+		return EINVAL;
+
+	if (prm->fmt != AUFMT_S16LE) {
+		warning("augain: format not supported (%s)\n",
+			aufmt_name(prm->fmt));
+		return ENOTSUP;
+	}
+
+	st = mem_zalloc(sizeof(*st), enc_destructor);
+	if (!st)
+		return EINVAL;
+
+	*stp = (struct aufilt_enc_st *)st;
+
+	return 0;
+}
+
+
+static int encode_frame(struct aufilt_enc_st *st, struct auframe *af)
+{
+	unsigned int abs_sample, highest_abs_sample = 0;
+	double encode_gain = gain;
+	double highest_possible_gain;
+	int16_t sample, gained_sample;
+
+	if (!st || !af || !af->sampv || !af->sampc)
+		return EINVAL;
+
+	for (size_t i=0; i<af->sampc; i++) {
+		abs_sample = abs(((int16_t *)(af->sampv))[i]);
+		if (abs_sample > highest_abs_sample)
+			highest_abs_sample = abs_sample;
+	}
+
+	highest_possible_gain = 32767.0 / highest_abs_sample;
+	if (encode_gain > highest_possible_gain)
+		encode_gain = highest_possible_gain;
```

Comment:
It looks like the real gain is different for each frame, depending on the frame energy.
Does this algorithm work, is there any audible noise or artefacts ?

Reply:
Yes.  I thought that it is better than cutting those sample values that that would not fit 32768 - 32767 range.

I have made tests and haven't noticed anything very bad.  It is trade of between not hearing anything and hearing, but with some artifacts.

Reply:
I made several tests using both adaptive and fixed algorithm and noticed less artifacts with the adaptive one.  Thus I don't think  that it is worth to add fixed algorithm to this PR.  It can be added later if someone needs it.

From my point of view, this PR is done.